src/wallet/coinselection.cpp

```diff
@@ -0,0 +1,174 @@
+{
+    out_set.clear();
+    CAmount value_track = 0;
+
+    if (utxo_pool.empty()) {
+        return false;
+    }
+
+    int depth = 0;
+    int remaining_tries = 100000;
+    std::vector<bool> selection; // select the utxo at this index
+    selection.assign(utxo_pool.size(), false);
+    bool done = false;
+    bool backtrack = false;
+
+    // Sort the utxo_pool
+    std::sort(utxo_pool.begin(), utxo_pool.end(), descending);
+
+    // Calculate lookahead
+    CAmount lookahead = 0;
+    for (const CInputCoin& utxo : utxo_pool) {
+        lookahead += utxo.txout.nValue;
+    }
+    if (lookahead < target_value) {
+        return false;
+    }
+
+    // Best solution
+    CAmount curr_waste = 0;
+    std::vector<bool> best_selection;
+    CAmount best_waste = MAX_MONEY;
+
+    // Depth First search loop for choosing the UTXOs
+    while (!done)
+    {
+        // Failure mode
+        if (remaining_tries <= 0) { // Too many tries, exit
+            break;
+        }
+        // Conditions for starting a backtrack
+        else if ( value_track + lookahead < target_value || // Cannot possibly reach target with the amount remaining in the lookahead.
+                    value_track > target_value + cost_of_change || // Selected value is out of range, go back and try other branch
+                    (curr_waste > best_waste && (utxo_pool.at(0).fee - utxo_pool.at(0).long_term_fee) > 0)) { // Don't select things which we know will be more wasteful if the waste is increasing
+            backtrack = true;
+        } else if (value_track >= target_value) { // Selected value is within range
+            curr_waste += (value_track - target_value); // This is the excess value which is added to the waste for the below comparison
+            // Adding another UTXO after this check could bring the waste down if the long term fee is higher than the current fee.
+            // However we are not going to explore that because this optimization for the waste is only done when we have hit our target
+            // value. Adding any more UTXOs will be just burning the UTXO; it will go entirely to fees. Thus we aren't going to
+            // explore any more UTXOs to avoid burning money like that.
+            if (curr_waste <= best_waste) {
+                best_selection.assign(selection.begin(), selection.end());
+                best_waste = curr_waste;
+            }
+            curr_waste -= (value_track - target_value); // Remove the excess value as we will be selecting different coins now
+            backtrack = true;
+        }
+
+        // Backtracking, moving backwards
+        if (backtrack) {
+            backtrack = false; // Reset
+            --depth;
+
+            // Walk backwards to find the first utxo which has not has its second branch traversed
+            while (!selection.at(depth)) {
```

Comment:
I'm not loving the way we track depth, it gets a bit confusing to make sure it stays in bounds.  I don't know if it's possible to hit in practice, but if you are trying to pay 0 total value, then you'll end up here trying to do `selection.at(-1)`.

Reply:
I agree, but we shouldn't have a 0 value target in practice. I could just add a check to make sure that `depth` is greater than 0 before decrementing it.

Reply:
yeah we should do something.
maybe change while loop to:
`while (depth >= 0 && !selection.at(depth)) {`
and then take the `if (depth < 0)` check out of the while loop and put it afterwards, and then it's 
```
if (depth < 0) {
    done = true;
}
else { 
...
}
```

and then you can get rid of the `if (!done)`
and in fact you could actually break instead of setting done=true and get rid of done and have the while loop be a for loop over remaining tries instead